From db0bc8b315c543272259c2cef075bca6ba3bb2b5 Mon Sep 17 00:00:00 2001
From: Srinu Jella <quic_sjella@quicinc.com>
Date: Mon, 18 Jul 2022 16:30:10 +0530
Subject: [PATCH 17/50] BT app: Add SHO support for split a2dp sink

- Add support for soft handoff for split
  a2dp sink cases.

CRs-Fixed: 3249572
Change-Id: If6a85b797c7e075ea52dd7e776e2e207a3069da6
---
 .../bluetooth/a2dpsink/A2dpSinkService.java   | 69 +++++++++++++++----
 .../android/bluetooth/btservice/Config.java   |  1 +
 2 files changed, 58 insertions(+), 12 deletions(-)

diff --git a/src/com/android/bluetooth/a2dpsink/A2dpSinkService.java b/src/com/android/bluetooth/a2dpsink/A2dpSinkService.java
index 0d10ef9cc..5e7b281f5 100644
--- a/src/com/android/bluetooth/a2dpsink/A2dpSinkService.java
+++ b/src/com/android/bluetooth/a2dpsink/A2dpSinkService.java
@@ -70,12 +70,15 @@ public class A2dpSinkService extends ProfileService {
     private static A2dpSinkStreamHandler mA2dpSinkStreamHandler;
     private static A2dpSinkService sService;
     protected static BluetoothDevice mStreamingDevice;
+    protected static BluetoothDevice mHandOffPendingDevice = null;
 
     private static int mMaxA2dpSinkConnections = 1;
     private A2dpSinkVendorService mA2dpSinkVendor;
     private AudioManager mAudioManager;
     public static final int MAX_ALLOWED_SINK_CONNECTIONS = 2;
     private static boolean sAudioIsEnabled = false;
+    private static boolean sIsHandOffPending = false;
+    private static boolean mIsSplitSink = false;
     static {
         classInitNative();
     }
@@ -91,6 +94,8 @@ public class A2dpSinkService extends ProfileService {
         mMaxA2dpSinkConnections = Math.min(
                 SystemProperties.getInt("persist.vendor.bt.a2dp.sink_conn", 1),
                 MAX_ALLOWED_SINK_CONNECTIONS);
+        mIsSplitSink = SystemProperties.
+          getBoolean("persist.vendor.bluetooth.split_a2dp_sink", false);
         mAudioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
         mA2dpSinkVendor = new A2dpSinkVendorService(this);
         if (mA2dpSinkVendor != null) {
@@ -384,7 +389,9 @@ public class A2dpSinkService extends ProfileService {
                 initiateHandoffOperations(device);
                 if (mStreamingDevice != null && !mStreamingDevice.equals(device)) {
                     Log.d(TAG, "updating streaming device after avrcp status command");
-                    mStreamingDevice = device;
+                    if(!mIsSplitSink) {
+                        mStreamingDevice = device;
+                    }
                 }
                 //mStateMachine.sendMessage(A2dpSinkStateMachine.EVENT_AVRCP_TG_PLAY);
                 mA2dpSinkStreamHandler.obtainMessage(
@@ -634,7 +641,7 @@ public class A2dpSinkService extends ProfileService {
                 Log.d(TAG, "current connected device: " + device + "is different from previous device");
                 initiateHandoffOperations(device);
                 mStreamingDevice = device;
-            } else if (device != null) {
+            } else if (device != null && !mIsSplitSink) {
                 mStreamingDevice = device;
             }
             if (mAudioManager != null) {
@@ -650,20 +657,41 @@ public class A2dpSinkService extends ProfileService {
     private void onAudioStateChanged(byte[] address, int state) {
         BluetoothDevice device = getDevice(address);
         Log.d(TAG, "onAudioStateChanged. Audio State = " + state + ", device:" + device);
-
+        Log.d(TAG, "onAudioStateChanged. mStreamingDevice = " + mStreamingDevice);
         A2dpSinkStateMachine stateMachine = mDeviceStateMap.get(device);
         if (stateMachine == null) {
+            Log.d(TAG, "onAudioStateChanged return");
             return;
         }
 
         if (state == StackEvent.AUDIO_STATE_STARTED) {
+            if(sIsHandOffPending == true) {
+                mStreamingDevice = mHandOffPendingDevice;
+                sIsHandOffPending = false;
+                mHandOffPendingDevice = null;
+            }
             initiateHandoffOperations(device);
             mStreamingDevice = device;
             mA2dpSinkStreamHandler.obtainMessage(
                     A2dpSinkStreamHandler.SRC_STR_START).sendToTarget();
-        } else if (state == StackEvent.AUDIO_STATE_STOPPED) {
+
+        } else if (state == StackEvent.AUDIO_STATE_STOPPED ||
+                   state == StackEvent.AUDIO_STATE_REMOTE_SUSPEND) {
             mA2dpSinkStreamHandler.obtainMessage(
                     A2dpSinkStreamHandler.SRC_STR_STOP).sendToTarget();
+            if (sIsHandOffPending == true &&
+                mStreamingDevice != null && mHandOffPendingDevice != null) {
+                Log.d(TAG, "current stopped device: " + device);
+                mA2dpSinkStreamHandler.obtainMessage(
+                      A2dpSinkStreamHandler.START_SINK).sendToTarget();
+                sAudioIsEnabled = true;
+                return;
+            }
+
+            if(mIsSplitSink) {
+                sAudioIsEnabled = false;
+                mStreamingDevice = null;
+            }
         }
     }
 
@@ -679,17 +707,34 @@ public class A2dpSinkService extends ProfileService {
     }
 
     public void onStartIndCallback(byte[] address) {
-        Log.d(TAG, "onStartIndCallback" );
-       if(sAudioIsEnabled == false) {
-         mA2dpSinkStreamHandler.obtainMessage(
-             A2dpSinkStreamHandler.START_SINK).sendToTarget();
-         sAudioIsEnabled = true;
-       }
+        BluetoothDevice device = getDevice(address);
+        Log.d(TAG, "onStartIndCallback dev " + device + "streaming device" + mStreamingDevice);
+        if (mStreamingDevice != null && !mStreamingDevice.equals(device)) {
+            Log.d(TAG, "current streaming device: " + mStreamingDevice);
+            if(sIsHandOffPending == true &&
+               mHandOffPendingDevice.equals(device)) {
+              Log.d(TAG, "handoff already triggered: " + mStreamingDevice);
+              return;
+            }
+            sIsHandOffPending = true;
+            mHandOffPendingDevice = device;
+            initiateHandoffOperations(device);
+            mA2dpSinkStreamHandler.obtainMessage(
+                A2dpSinkStreamHandler.STOP_SINK).sendToTarget();
+            sAudioIsEnabled = false;
+            return;
+        }
+
+        if(sAudioIsEnabled == false) {
+             mA2dpSinkStreamHandler.obtainMessage(
+                 A2dpSinkStreamHandler.START_SINK).sendToTarget();
+            sAudioIsEnabled = true;
+        }
     }
 
     public void onSuspendIndCallback(byte[] address) {
-        // TODO to call set param to intimate Audio HAL
-        Log.d(TAG, "onSuspendIndCallback" );
+        BluetoothDevice device = getDevice(address);
+        Log.d(TAG, "onSuspendIndCallback" + device);
         if(sAudioIsEnabled == true) {
           mA2dpSinkStreamHandler.obtainMessage(
               A2dpSinkStreamHandler.STOP_SINK).sendToTarget();
diff --git a/src/com/android/bluetooth/btservice/Config.java b/src/com/android/bluetooth/btservice/Config.java
index 7a8d1fa00..a2f4a83b8 100644
--- a/src/com/android/bluetooth/btservice/Config.java
+++ b/src/com/android/bluetooth/btservice/Config.java
@@ -450,6 +450,7 @@ public class Config {
 
         if (mIsSplitSink) {
            SystemProperties.set("persist.vendor.service.bt.a2dp.sink", "true");
+           SystemProperties.set("persist.vendor.bt.a2dp.sink_conn", "2");
         }
 
         mIsA2dpSink = SystemProperties.getBoolean("persist.vendor.service.bt.a2dp.sink", false);
-- 
2.17.1

